packages/task/src/node/task-server.ts

```diff
@@ -0,0 +1,258 @@
+        @inject(TaskManager) protected readonly taskManager: TaskManager,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(TaskFactory) protected readonly taskFactory: TaskFactory
+    ) {
+        const taskManagerListener = taskManager.onDelete(id => {
+            const toDispose = this.taskToDispose.get(id);
+            if (toDispose !== undefined) {
+                toDispose.dispose();
+                this.taskToDispose.delete(id);
+            }
+        });
+        this.toDispose.push(taskManagerListener);
+    }
+
+    dispose() {
+        // do nothing
+    }
+
+    disconnectClient(client: TaskClient) {
+        const idx = this.clients.indexOf(client);
+        if (idx > -1) {
+            this.clients.splice(idx, 1);
+        }
+    }
+
+    getRunningTasks(context?: string | undefined): Promise<TaskInfo[]> {
+        const taskinfo: TaskInfo[] = [];
+
+        const tasks = this.taskManager.getTasks(context);
+        if (tasks !== undefined) {
+            for (const task of tasks) {
+                taskinfo.push(
+                    {
+                        taskId: task.id,
+                        terminalId: (task.processType === 'terminal') ? task.process.id : undefined,
+                        processId: (task.processType === 'raw') ? task.process.id : undefined,
+                        osProcessId: task.process.pid,
+                        command: task.command,
+                        label: task.label
+                    });
+            }
+        }
+        return Promise.resolve(taskinfo);
+    }
+
+    run(task: TaskOptions, ctx?: string): Promise<TaskInfo> {
+        return this.doRun(task, ctx);
+    }
+
+    protected doRun(options: TaskOptions, ctx?: string): Promise<TaskInfo> {
+        return new Promise<TaskInfo>(async (resolve, reject) => {
+            // on windows, prefer windows-specific options, if available
+            const processOptions = (isWindows && options.windowsProcessOptions !== undefined) ?
+                options.windowsProcessOptions : options.processOptions;
+
+            const command = processOptions.command;
+            const cwd = FileUri.fsPath(options.cwd);
+
+            // Use task's cwd with spawned process and pass node env object to
+            // new process, so e.g. we can re-use the system path
+            processOptions.options = {
+                cwd: cwd,
+                env: process.env
+            };
+
+            // When we create a process to execute a command, it's difficult to know if it failed
+            // because the executable or script was not found, or if it was found, ran, and exited
+            // unsuccessfully. So here we look to see if it seems we can find a file of that name that
+            // is likely to be the one we want, before attemting to execute it.
+            this.findCommand(command, cwd)
```

Comment:
why do we do `findCommand`? Aren't the error message from the process factories good?

Reply:
no, not so much. It's difficult to distinguish between the cases where a script or binary is found and executed, with an abnormal exit, versus the script/binary not being found at all (I think because of limitations in node-pty). Maybe we should enhance node-pty?